### ISSUE TITLE
Last layer activation

### ISSUE THREAD

**Issue by miroblog** [sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning#21]
High, thank you for making such a wonderful tutorial! It helped with my study greatly.

I see that, the linear layer is used to compute the score of each vocab. 
as in ` self.fc = nn.Linear(decoder_dim, vocab_size)` . 

In other paper, they would use softmax activation, since we are training with 
`nn.CrossEntropyLoss().to(device)`  

what are your thoughts on this? Will there be performance increase when linear activation is swapped with softmax?  (of course...softmax is computationally expensive with larger vocab size)

**Comment by sgrvinod** [comment]
Hi @miroblog, thank you, glad it's helpful for you.

[We're training with the Cross Entropy loss too](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning/blob/215f86e0518c828f0efeb9e64d0d197f4badd119/train.py#L86), so the softmax activation _has_ to be applied. However, this is done internally by [torch.nn.CrossEntropyLoss()](https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss). It's mentioned in its documentation that -

> This criterion combines nn.LogSoftmax() and nn.NLLLoss() in one single class.

> The input is expected to contain scores for each class.

These are just raw, unactivated scores. Therefore, we do not need to apply the softmax (or log) operations ourselves.

On the other hand, if we use [torch.nn.NLLLoss()](https://pytorch.org/docs/stable/nn.html#torch.nn.NLLLoss), which is generally the same as the Cross Entropy Loss, we need to apply both the log and softmax operations ourselves.

> The input given through a forward call is expected to contain log-probabilities of each class.

**Comment by miroblog** [comment]
Oh thank you for your prompt response :)
